### ISSUE TITLE
Support Custom Protocols

### ISSUE THREAD

**Issue by lidel** [lidel/ipfs-firefox-addon#164]
WebExtension  API is under development:

- https://bugzilla.mozilla.org/show_bug.cgi?id=1271553
- https://bugzilla.mozilla.org/show_bug.cgi?id=1296885 (closed duplicate)

**Comment by lidel** [comment]
A simplified handler support landed in Firefox 54:
- https://bugzilla.mozilla.org/show_bug.cgi?id=1310427 

This a a thin UX on top of `Navigator.registerProtocolHandler()`:
- https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler

**Comment by lidel** [comment]
It seems that this new API has the same naming limitation as Chrome, firefox-54.0a2 says:

> Reading manifest: Error processing protocol_handlers.0.protocol: Value must either: be one of ["bitcoin", "geo", "im", "irc", "ircs", "magnet", "mailto", "mms", "news", "nntp", "sip", "sms", "smsto", "ssh", "tel", "urn", "webcal", "wtai", "xmpp"], or **match the pattern /^(ext|web)\+[a-z0-9.+-]+$/**

**Comment by Kubuxu** [comment]
When we get official registration of the schema we should be able to request the exemption to be added.

**Comment by lidel** [comment]
I've just tested a _ridiculous_ PoC that restores support for `ipfs://` `fs://` etc in WebExtension:

- We know that: 
  - permission `<all_urls>` fires `onBeforeRequest` for every URL
  - when URL starting with _an unknown protocol_ is entered in location bar, a browser url-escapes entire address and converts it to a search query, e.g:
   `ipfs://QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR` → `https://www.google.com/search?q=ipfs%3A%2F%2FQmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR&ie=utf-8&oe=utf-8&client=firefox-b-ab`
- So what happens if you:
  - detect requests with URL containing url-encoded `://` ( `%3A%2F%2F`)
  - find ones that match a query starting with one of custom protocols: `/=(ipfs|ipns|fs|dweb)%3A%2F(%2F[^&]+)/` 
  - extract IPFS path and check if passes `IsIpfs.path` test
    - if so, replace request to search engine with request to resource at public gateway  `'https://ipfs.io/' + extractedIpfsPath` 
    - if not, return original request

Well.. it just works 🚀  🙃 (not sure is I am proud or ashamed, probably both)

Expect PR in near future. My plan is to polish and ship this workaround as default behavior, so that there is no functional regression when users update from v1.6.0 to v2.x.x.

Of course there will be a preference that disables this: to pass review at Mozilla and in case someone really wants to search for `ipfs://QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR`

**Comment by lidel** [comment]
Custom Protocols in WebExtension – status update: 

_Poor Man's Protocol Handlers_ (https://github.com/ipfs/ipfs-companion/issues/164#issuecomment-328374052)  are enabled by default in [v2.0.10](https://github.com/ipfs/ipfs-companion/releases/tag/v2.0.10).
- Due to internalized shame 🙃 I've put it under this innocent label: 
  > ![2017-09-28-014201_525x78_scrot](https://user-images.githubusercontent.com/157609/30942978-4b4873d2-a3ee-11e7-81a8-5c6f69831d14.png)

- [The hack](https://github.com/ipfs/ipfs-companion/issues/164#issuecomment-328374052) works in both Firefox and Chrome and is the best thing we can do as of late 2017.
- I've addressed [feedback](https://github.com/ipfs/ipfs-companion/pull/283#issuecomment-330006486) from @lgierth  to support only URL schemes: `ipfs://`, `ipns://` and the URI `dweb:` (`fs:` is no longer supported)
- FYSA there is a small possibility that we will be able to modify Location bar in Firefox at some point, but that depends on the direction of [[Bug 1394089] Support Advanced Locationbar as a webextension or make new omnibox APIs](https://bugzilla.mozilla.org/show_bug.cgi?id=1394089).

**Comment by dag** [comment]
The `ipfs` scheme has no "authority" as per the RFC so it should not include the double forward slashes. It is debatable whether the `ipns` scheme includes an authority; it includes a domain name, but it's all routed through IPFS and for example there is no port number and the domain isn't necessarily a peer. Thus even with IPNS the domain is merely an identifier and not an "authority". I therefore propose that it would be more in compliance with the specification to have:

* `ipfs:{hash}` and
* `ipns:{domain}`

Double slashes are a wart of the URI spec, so it's nice to be able to avoid them. The spec defines authority thus:

```
authority   = [ userinfo "@" ] host [ ":" port ]
```

IPFS is an unauthenticated peer-to-peer network, so there is no userinfo and no singular host or port.

### Why I could be wrong

I don't think I'm wrong for the `ipfs` scheme, but I'm less certain for the `ipns` scheme. It could be argued that when the RFC defines `host` to include "registered name", it is very vague as to what that means and it could apply to IPNS. But I think most people would expect that if there is two slashes followed by a domain name, the computer will connect to that domain name, and it could be followed by a port number. Neither of those are necessarily or at all true with IPNS.

**Comment by lidel** [comment]
@dag  Thanks! Mind moving it to  https://github.com/ipfs/specs/pull/152? It is a better place for  this discussion, more people are watching that repo.  This issue is only about implementing current consensus in the browser extension :)

(If you are interested in **tl;dr** with rationale behind use of URL for `ipfs://` and `ipns://` and URI for `dweb:`, see ["Four stages of the upgrade path for path addressing"](https://github.com/ipfs/specs/pull/152#issuecomment-284628862))

**Comment by lidel** [comment]
As a result of a [thread at dev-addons](https://mail.mozilla.org/pipermail/dev-addons/2018-January/003553.html) Andre opened a ticket to [Whitelist custom protocol handlers for work on decentralization technologies with WebExtensions](https://bugzilla.mozilla.org/show_bug.cgi?id=1428446).
If accepted, Firefox users will no longer have to rely on hack from https://github.com/ipfs-shipyard/ipfs-companion/issues/164#issuecomment-328374052, but will get  UX from https://github.com/ipfs-shipyard/ipfs-companion/issues/164#issuecomment-282513891 without `web+` prefix. 
A small step, but worth mentioning here, as it would solve #348 :)

**Comment by soapdog** [comment]
@lidel it worked! The patch landed on nightly (I think, I am new to this process). Check out https://hg.mozilla.org/mozilla-central/rev/c2cb8a06bcf1

(PS: I never used IPFS but I thought that since I was going through the process of doing a patch to whitelist scuttlebutt, I should include ipfs and dat as well... glad it worked)

Also I was checking a bit from this thread, and (please correct if I am wrong) apparently you're intercepting a protocol and doing a web redirect. Is that right? If so, then I believe you might be able to use the WebRequest API from WebExtensions to intercept the request and fiddle with the origin handlers you wanted, it might help there.

**Comment by lidel** [comment]
@soapdog that is really cool, thank you!!  🎉  ❤️ 
We'll switch to non-prefixed handler as soon it lands in regular Firefox release. 👌 

Yes, we are intercepting and redirecting. The problem is  that we want to control Origin _in the browser itself_, to isolate DApps from each other. AFAIK modifying `Origin` header in requests will only spoof Origin for remote server, while things like JS on a page will still use standard Origin for cookies and storage.  It won't work without [webextension support for synthetic origins/protocols](https://bugzilla.mozilla.org/show_bug.cgi?id=1271553).

**Comment by mitra42** [comment]
I was one of the authors on some drafts of some of those specs (URN and URI) so I can give you some of the thinking that went behind decisions.    

Double slashes were in there to distinguish them from a single leading slash.  i.e /foo means go to the same protocol, same host, and ask for /foo. While //xx.yy.zz meant to go the xx.yy.zz host first and ask for /foo.  That was a feature of all known protocols at that time and as pointed out doesn't mean anything in a host-less world. 

urn: was in there for two reasons (instead of isbn) 
a) because the part before the : implied a protocol, what you use to talk to the resolver, in practice it told you what library to hand the query off to. 
c) there is no isbn protocol, so it shouldn't really be before the first :, at the time there was a good likelyhood of there being a protocol for resolving URNs. 
c) because its a hard place for extensibility. Netscape added urn: around 1996 on my request to allow us to experiment, but there was no way they were going to add one protocol for each of isbn, and each of the other naming authorities, especially without any library/protocol to hand it off to.

**Comment by lidel** [comment]
FYSA some recent developments (Q3 2018):

- **Experimental Protocol Handler AP**I for WebExtensions is being designed as a part of [mozilla/libdweb](https://github.com/mozilla/libdweb/): 
  >  The Protocol API allows you to handle custom protocols from your Firefox extension. This is different from the existing WebExtensions protocol handler API in that it does not register a website for handling corresponding URLs but rather allows your WebExtension to implement the handler.
  > More: https://github.com/mozilla/libdweb/#protocol-api

  - Still experimental, requires a special build, but the goal is for this to land in regular builds of Firefox  Nightly at some point in the future :sparkles: 

  - We've built a PoC handler that loads data over raw IPFS and keeps `ipfs://` in address bar.
   It can be found in `libdweb` branch.  More info in PR https://github.com/ipfs-shipyard/ipfs-companion/pull/533  and [/libdweb/docs/libdweb.md](https://github.com/ipfs-shipyard/ipfs-companion/blob/libdweb/docs/libdweb.md)
    

- **Safelisting DWeb Protocols**  | https://github.com/ipfs-shipyard/arewedistributedyet/issues/23
  - **tl;dr** The goal is to be able to register it in all browsers without `web+` prefix.   
     (Chrome 67 requires `web+` prefix for non-safelisted protocols)
  - Firefox already allows non-prefixed version, Chrome published [intent-to-implement](https://www.chromestatus.com/feature/4776602869170176).

**Comment by lidel** [comment]
If DWeb protocools get safelisted, we want feature parity with Firefox to solve UX issues mentioned in https://github.com/arewedistributedyet/arewedistributedyet/issues/23#issuecomment-412892777

Filled a Chrome bug: [Extensions API should implement manifest.json/protocol_handlers](https://bugs.chromium.org/p/chromium/issues/detail?id=883274).

**Comment by ingokeck** [comment]
According to the URI syntax in RFC3986, it should not be "ipfs://{cidv1b32}", instead it should be "ipfs:{cidv1b32}". Same for {hash}. The extention should then use its default IPFS instance to retrieve the object {cidv1b32}. See [https://tools.ietf.org/html/rfc3986#section-3](https://tools.ietf.org/html/rfc3986#section-3)

'//' denotes a specific instance. So an IPFS service on localhost could be addressed with "ipfs://localhost/{cidv1b32}", while IPFS running on ipfs.io would be "ipfs://ipfs.io/{cidv1b32}"

**Comment by lidel** [comment]
Those are good points, thanks for raising them @ingokeck 

- :green_heart: For the sake of UX, I believe both should work.  For example, you can use `ipfs:bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi` in Firefox today and if you have IPFS Companion installed, it will resolve just fine. If we flip the default, the reverse should be true as well.

- :yellow_heart:  Removal of `//` needs discussion with Brave and Opera, but I suspect the only reason for `//` is technical debt in web browsers and being pragmatic. Browsers are HTTP-centric and existing implementation leverage this property of RFC3986:

  > Non-validating parsers (those that merely separate a URI reference into
   its major components) will often ignore the subcomponent structure of
   authority, treating it as an opaque string from the double-slash to
   the first terminating delimiter, until such time as the URI is dereferenced.

  For example Brave v1.19 uses CID as "authority" component and build Origin based on it. This provides us with security sandbox per content root (CID), but has a side-effect of `//` in the address bar. 

  **TLDR** Origin isolation is way more important than being visually compliant with RFC, but as long the Origin isolation per CID is maintained, we should be ok with removal of `//` 

- :broken_heart:  `ipfs://localhost/{cidv1b32}` is incompatible with our security model given how addressing is implemented in browsers
   - Each CID needs to create own Origin, and this will put all CIDs under single Origin, so a big NO. 
   -  Yes, this could be special-cased, but is very unlikely any vendor will want to touch critical code paths for the sake of cosmetics. In the past we looked into something simpler with Suborigins and even that did not go anywhere: https://github.com/ipfs/in-web-browsers/issues/66

**Comment by lidel** [comment]
5 years [after Firefox](https://github.com/ipfs/ipfs-companion/issues/164#issuecomment-282513891), Google Chrome is now willing to implement the [protocol_handler](https://developer.mozilla.org/en-US/docs/Web/Manifest/protocol_handlers) for registering custom handlers via web extensions: https://github.com/w3c/webextensions/issues/317#issuecomment-1472508987

**Comment by lidel** [comment]
We are still waiting for Chromium to allow registration with [protocol_handlers](https://developer.mozilla.org/en-US/docs/Web/Manifest/protocol_handlers) in browser extension manifest.

Once we have that, the next step will be to allow using the API for pointing protocol scheme at Service Worker, instead of URL. This modern approach requires changes tracked in:
- https://github.com/ipfs/in-web-browsers/issues/212
